File: 2015/22.py
from Read import *
import re
import math
import itertools
import functools

# Part 1


def simulate_fight(strategy, hp, mana, boss_hp, boss_dmg):
    spent_mana = 0
    armor = [0, 0]
    poison = [0, 0]
    recharge = [0, 0]

    for turn in range(len(strategy)):
        # Our turn

        hp -= 1  # Part 2
        if hp <= 0:
            return False

        # Tick statuses
        if armor[1] <= 0:
            armor[0] = 0
            armor[1] = 0
        armor[1] -= 1

        if poison[1] <= 0:
            poison[0] = 0
            poison[1] = 0
        else:
            boss_hp -= poison[0]
        poison[1] -= 1

        if recharge[1] <= 0:
            recharge[0] = 0
            recharge[1] = 0
        else:
            mana += recharge[0]
        recharge[1] -= 1

        if boss_hp <= 0:
            return spent_mana

        if strategy[turn] == 'missile':
            if mana < 53:
                return False
            
            spent_mana += 53
            mana -= 53
            boss_hp -= 4

        elif strategy[turn] == 'drain':
            if mana < 73:
                return False

            spent_mana += 73
            mana -= 73
            boss_hp -= 2
            hp += 2

        elif strategy[turn] == 'shield':
            if mana < 113:
                return False

            spent_mana += 113
            mana -= 113
            armor = [7, 6]

        elif strategy[turn] == 'poison':
            if mana < 173:
                return False

            spent_mana += 173
            mana -= 173
            poison = [3, 6]

        elif strategy[turn] == 'recharge':
            if mana < 229:
                return False

            spent_mana += 229
            mana -= 229
            recharge = [101, 5]

        # Boss's turn

        # Tick statuses
        if armor[1] <= 0:
            armor[0] = 0
            armor[1] = 0
        armor[1] -= 1

        if poison[1] <= 0:
            poison[0] = 0
            poison[1] = 0
        else:
            boss_hp -= poison[0]
        poison[1] -= 1

        if recharge[1] <= 0:
            recharge[0] = 0
            recharge[1] = 0
        else:
            mana += recharge[0]
        recharge[1] -= 1

        if boss_hp <= 0:
            return spent_mana

        hp -= max(1, boss_dmg - armor[0])
        if hp <= 0:
            return False

    return False


spells = {'missile': 53,
        'drain': 73,
        'shield': 113,
        'poison': 173,
        'recharge': 229}

# print(simulate_fight(['recharge', 'shield', 'drain', 'poison', 'missile'], 10, 250, 14, 8))

strategies = list(itertools.product(spells.keys(), repeat=9))
min_mana = 99999
for strategy in strategies:
    mana = simulate_fight(strategy, 50, 500, 55, 8)
    if 0 < mana < min_mana:
        min_mana = mana
        print(mana)

# 20:18 (2nd)

# Part 2

# 21:43 (1st)

File: v2/2015/22.R

# Load --------------------------------------------------------------------

rm(list = ls())
source("utilities.R")

# Problem -----------------------------------------------------------------

inp <- strsplit("", "\n")[[1]]  # Example
inp <- get_input("2015/22", parse = F, user = "blakeebets", cache = T)



# BFS

boss_hp <- 58
boss_dmg <- 9

hp <- 50
mana <- 500

cost <- c("missile" = 53, "drain" = 73, "shield" = 113, "poison" = 173, "recharge" = 229)

Q <- list(list(hp = hp, mana = mana, armor = 0, spent = 0, boss_hp = boss_hp, boss_dmg = boss_dmg, 
               effects = c("shield" = 0, "poison" = 0, "recharge" = 0)))

min_mana <- Inf

while(length(Q) > 0) {
    state <- Q[[1]]
    Q <- Q[-1]
    
    if(state[["boss_hp"]] < 0) {
        if(state[["spent"]] < min_mana) {
            min_mana <- state[["spent"]]    
            print(min_mana)
        }
        next
    } else if(state[["hp"]] <= 1) {  # Part 2: drain
        next
    } else if(state[["spent"]] > min_mana) {
        next
    }
    
    effects <- state[["effects"]]
    state[["hp"]] <- state[["hp"]] - 1  # Part 2: drain
    
    # Player turn
    
    # Apply effects to us
    if(effects[["poison"]] > 0) {
        state[["boss_hp"]] <- state[["boss_hp"]] - 3
    }
    
    if(effects[["recharge"]] > 0) {
        state[["mana"]] <- state[["mana"]] + 101
    }
    
    effects <- setNames(pmax(0, effects - 1), names(effects))
    state[["effects"]] <- copy(effects)
    
    # Try casting each spells
    new_states <- list()
    
    for(spell in names(cost)) {
        if(state[["mana"]] >= cost[[spell]]) {
            if(!(spell %in% names(effects)) || effects[[spell]] == 0) {
                if(spell == "missile") {
                    new_state <- copy(state)
                    
                    new_state[["boss_hp"]] <- new_state[["boss_hp"]] - 4
                    new_state[["mana"]] <- new_state[["mana"]] - cost[[spell]]
                    new_state[["spent"]] <- new_state[["spent"]] + cost[[spell]]
                    
                    new_states[[length(new_states) + 1]] <- new_state
                } else if(spell == "drain") {
                    new_state <- copy(state)
                    
                    new_state[["hp"]] <- new_state[["hp"]] + 2
                    new_state[["boss_hp"]] <- new_state[["boss_hp"]] - 2
                    new_state[["mana"]] <- new_state[["mana"]] - cost[[spell]]
                    new_state[["spent"]] <- new_state[["spent"]] + cost[[spell]]
                    
                    new_states[[length(new_states) + 1]] <- new_state
                } else if(spell == "shield") {
                    new_state <- copy(state)
                    
                    new_state[["mana"]] <- new_state[["mana"]] - cost[[spell]]
                    new_state[["spent"]] <- new_state[["spent"]] + cost[[spell]]
                    new_state[["effects"]][["shield"]] <- 6
                    
                    new_states[[length(new_states) + 1]] <- new_state
                } else if(spell == "poison") {
                    new_state <- copy(state)
                    
                    new_state[["mana"]] <- new_state[["mana"]] - cost[[spell]]
                    new_state[["spent"]] <- new_state[["spent"]] + cost[[spell]]
                    new_state[["effects"]][["poison"]] <- 6
                    
                    new_states[[length(new_states) + 1]] <- new_state
                } else if(spell == "recharge") {
                    new_state <- copy(state)
                    
                    new_state[["mana"]] <- new_state[["mana"]] - cost[[spell]]
                    new_state[["spent"]] <- new_state[["spent"]] + cost[[spell]]
                    new_state[["effects"]][["recharge"]] <- 5
                    
                    new_states[[length(new_states) + 1]] <- new_state
                }
            }
        }
    }
    
    # Boss turn
    for(ns in new_states) {
        effects <- ns[["effects"]]
        
        # Apply effects to boss
        if(effects[["poison"]] > 0) {
            ns[["boss_hp"]] <- ns[["boss_hp"]] - 3
        }
        
        if(effects[["recharge"]] > 0) {
            ns[["mana"]] <- ns[["mana"]] + 101
        }
        
        if(effects[["shield"]] > 0) {
            ns[["armor"]] <- 7
        } else {
            ns[["armor"]] <- 0
        }
        
        ns[["effects"]] <- setNames(pmax(ns[["effects"]] - 1, 0), names(ns[["effects"]]))
        
        effects <- pmax(0, effects - 1)
        
        # Deal damage
        ns[["hp"]] <- ns[["hp"]] - pmax(1, ns[["boss_dmg"]] - ns[["armor"]])
        
        Q <- c(Q, list(ns))
    }
    
    Q <- Q[order(sapply(Q, function(x) x[["hp"]] - x[["boss_hp"]]))]
}

# 25:08 (3rd)

# 27:12 (4th)

